### ISSUE TITLE
CUDA/Clang-16/C++2b: std::isfinite is incorrectly resolved as isfinite from __clang_cuda_cmath.h

### ISSUE THREAD

**Issue by andrewcorrigan** [llvm/llvm-project#61629]
I have a minimal reproducer below which fails with:
- `clang-16`, `-std=c++2b`

But, compiles with:
- `clang-16`, `-std=c++20`
- `clang-15`, `-std=c++20`
- `clang-15`, `-std=c++2b`

Minimal reproducer:

```cpp
#include <cuda.h>
#include <cuda_runtime_api.h>

#include <numeric>

int main(int argc, char** argv)
{
    auto test = [&](auto x)
    {
        return std::isfinite(x);
    };

    test(-1.0f);

    return 0;
}
```

This compiles in C++20 mode with Clang-16, as well as both C++20 and C++2b mode with Clang-15:
```
% /install/llvm-16.0.0/bin/clang++ -std=c++2b -x cuda test.cpp 
% /install/llvm-15.0.0/bin/clang++ -std=c++2b -x cuda test.cpp 
% /install/llvm-15.0.0/bin/clang++ -std=c++20 -x cuda test.cpp 
```

However, the reproducer above fails to compile in C++2b mode with Clang-16
```
% /install/llvm-16.0.0/bin/clang++ -std=c++2b -x cuda test.cpp 
test.cpp:10:21: error: reference to __device__ function 'isfinite' in __host__ __device__ function
        return std::isfinite(x);
                    ^
test.cpp:13:9: note: called by 'main'
    test(-1.0f);
        ^
/install/llvm-16.0.0/lib/clang/16/include/__clang_cuda_cmath.h:103:17: note: 'isfinite' declared here
__DEVICE__ bool isfinite(float __x) { return ::__finitef(__x); }
                ^
test.cpp:10:21: error: reference to __device__ function 'isfinite' in __host__ __device__ function
        return std::isfinite(x);
                    ^
test.cpp:14:9: note: called by 'main'
    test(-2.0);
        ^
/install/llvm-16.0.0/lib/clang/16/include/__clang_cuda_cmath.h:107:17: note: 'isfinite' declared here
__DEVICE__ bool isfinite(double __x) { return ::__isfinited(__x); }
                ^
2 errors generated when compiling for host.
```

**Comment by Artem-B** [comment]
The problem appears to be libc++-specific. -std=c++2b works fine with libstdc++: https://godbolt.org/z/a4ebf8Yfh

I suspect it has something to do with `isfinite` becoming a `constexpr` function in c++23: https://en.cppreference.com/w/cpp/numeric/math/isfinite

constexpr functions are treated as `__host__ __device__`, and clang headers are always providing a `__device__` implementation, which may have higher precedent and lead to the error you see.

However, given that libstdc++ still works suggests that it probably didn't make isfinite a constexpr yet, so the fix will have to depend on both C++ version  and the standard c++ library variant.

**Comment by Artem-B** [comment]
Interesting. libc++ does not define `__cpp_lib_constexpr_cmath`, yet at least some math functions *are* constexpr now with -std=c++2b. Making everything work for all C++ and stdlib versions is going to be tricky. 

A bigger problem is that while we can deal with the issues with the functions declared/defined in clang's own headers, there's little we can do about the functions provided by the CUDA headers themselves, and they also currently carry only non-constexpr math functions. What it means in practice is that c++23 will likely work only with some future CUDA version where the headers are in sync with what c++23 expects.

The bottom line for now is that c++23 will likely be broken with CUDA for a while.

**Comment by andrewcorrigan** [comment]
Thank you for explaining everything.  I wanted to mention two additional points: 

First, I included `<numeric>`, but I think it should have been `<cmath>`.  Sorry for confusing those, however, the error persists even changing the include. 

Second, when extracting a reproducer from my code, I noticed that the call to `std::isfinite` is only a problem **inside** of a lambda.   If called from outside, there's no issue.  So for example, the following compiles just fine.
```
#include <cuda.h>
#include <cuda_runtime_api.h>

#include <cmath>

int main(int argc, char** argv)
{
    auto a = std::isfinite(1.0f);

    return 0;
}
```

**Comment by Artem-B** [comment]
CUDA's relationship with constexpr and lambda is ... complicated.

For starters, `__host__`, `__device__` and `__host__ __device__` have a rather intricate impact on the function overload resolution:
https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaCUDA.cpp#L178

`constexpr` and `lambda` make it all much more interesting because they are treated as if they have implicit `__host__` `__device__` attributes. So, when a function in libc++ changes from  a regular (`__host__`) function to a `constexpr` (`__host__ __device__`) one, it does tend to break the carefully aranged parallel world of GPU-side function overloads provided by clang and CUDA headers. 

All you need to reproduce the original issue is some sort of `__host__ __device__` function calling `std::isfinite()`. It could be a constexpr function, a lambda, or a function with explicit attributes. E.g https://godbolt.org/z/54GGaMjfM

**Comment by andrewcorrigan** [comment]
Thank you for further explaining this.  Based on that I see a workaround to still allow for C++23 compilation by decorating the surrounding lambda with `__host__`.   https://godbolt.org/z/Y1qGj48nc

**Comment by Artem-B** [comment]
That would be one way to do it in this specific case, but it's not a very good general workaround. Granted, it's better than not being able to compile, but...

The problem is that the code that triggers the issue may happen to be located in the sources that are not under user's control. Your original reproducer is plain C++ code which has nothing to do with CUDA and could conceivably be encountered in some C++ library headers. In the libc++ itself, for example.

**Comment by miscco** [comment]
Yeah, device lambdas are ... special.

For what it is worth, there is a PR to implement constexpr complex operations so we are in the process of catching up https://github.com/NVIDIA/libcudacxx/pull/397

**Comment by Artem-B** [comment]
Thanks for the heads-up. That will likely run into the same problem and the switch to constexpr would need to be matched in clang's wrapper for `complex`. Ideally we should start incorporating limited CUDA support into libc++. In most cases we would just need to add `__host__ __device__` attributes on non-constexpr functions that are generic enough to compile on GPU. A small subset may need to remain `__host__` only and would either remain unimplemented on GPU, or we'll provide GPU-side implementation either in libc++ itself or in clang wrappers. I didn't have time to look at the specifics yet, but it looks like this may need to happen soon.
